allow named adapters and add receptionist to system

rkuhn opened the pull request:
This is the result of removing the process DSL from #22087, allowing that work to mature at [rkuhn/akka-typed-session](https://github.com/rkuhn/akka-typed-session).

rkuhn committed:
allow named adapters and add receptionist to system

rkuhn commented:
Oops, forgot to add tests—without the process DSL there are of course no users of the new feature.

ktoso commented on akka-actor/src/main/scala/akka/actor/dungeon/Children.scala:
```diff
@@ -59,8 +59,10 @@ private[akka] trait Children { this: ActorCell ⇒
-    val childPath = new ChildActorPath(self.path, randomName(new java.lang.StringBuilder("$$")), ActorCell.newUid())
+  private[akka] def addFunctionRef(f: (ActorRef, Any) ⇒ Unit, name: String = ""): FunctionRef = {
+    val r = randomName(new java.lang.StringBuilder("$$"))
+    val n = if (name != "") s"$r-$name" else r
```
ok

ktoso commented on akka-typed/src/main/scala/akka/typed/ActorContext.scala:
```diff
@@ -124,7 +124,7 @@ trait ActorContext[T] {
    * these ActorRefs or to stop them when no longer needed.
    */
-  def spawnAdapter[U](f: U ⇒ T): ActorRef[U]
+  def spawnAdapter[U](f: U ⇒ T, name: String = ""): ActorRef[U]
```
👍 
though i wonder if it would be safer to add them as overloads, in case we need to evolve these apis hmmm

rkuhn replied:
good point, and since it is in a trait doubly so; will change

ktoso reviewed (approved):
LGTM, though not fully getting which bit was the really needed change to allow Processes be externally defined?

rkuhn commented:
Oh, it was nothing that fundamental, apart from needing a Receptionist for anything serious. Naming the processes’ references is very useful for logging, debugging, etc.

ktoso commented:
Ok cool, LGTM, happy to get this in then.

ktoso commented:
Mima issues resolved via github but needs rebase anyway :~

johanandren commented on akka-typed/src/test/scala/akka/typed/ActorContextSpec.scala:
```diff
@@ -523,6 +522,15 @@ class ActorContextSpec extends TypedSpec(ConfigFactory.parseString(
       }
     })
+
+    def `41 must create a named adapter`(): Unit = sync(setup("ctx40") { (ctx, startWith) ⇒
```
ctx41 ?

rkuhn replied:
yep

johanandren reviewed (approved):
LGTM

rkuhn committed:
fix #21653

rkuhn commented:
Rewrote the default argument to an overload (for future BC benefits), fixed the ctx41 copy-pasta, and rebased on top of current MiMa exception. Shall I merge after validation?
